modules/howtos/examples/transactions_example.py

```diff
@@ -0,0 +1,449 @@
+import sys
+import traceback
+from typing import TYPE_CHECKING
+
+from couchbase.auth import PasswordAuthenticator
+from couchbase.cluster import Cluster
+from couchbase.durability import DurabilityLevel, ServerDurability
+from couchbase.exceptions import (DocumentNotFoundException,
+                                  TransactionCommitAmbiguous,
+                                  TransactionFailed)
+from couchbase.options import ClusterOptions, TransactionConfig
+from couchbase.transactions import TransactionQueryOptions
+
+if TYPE_CHECKING:
+    from couchbase.transactions import AttemptContext
+
+def main():
+    # tag::config[]
+    opts = ClusterOptions(authenticator=PasswordAuthenticator("Administrator", "password"),
+                          transaction_config=TransactionConfig(
+        durability=ServerDurability(DurabilityLevel.PERSIST_TO_MAJORITY)))
+
+    cluster = Cluster.connect('couchbase://localhost', opts)
+    # end::config[]
+
+    test_doc = "foo"
+
+    # tag::ts-bucket[]
+    # get a reference to our bucket
+    bucket = cluster.bucket("travel-sample")
+    # end::ts-bucket[]
+
+    # tag::ts-collection[]
+    # get a reference to our collection
+    collection = bucket.scope("inventory").collection("airline")
+    # end::ts-collection[]
+
+    # tag::ts-default-collection[]
+    # get a reference to the default collection, required for older Couchbase server versions
+    collection_default = bucket.default_collection()
+    # tag::ts-default-collection[]
+
+    # Set up for what we'll do below
+    remove_or_warn('doc-a')
+    remove_or_warn('doc-b')
+    remove_or_warn('doc-c')
+    remove_or_warn(test_doc)
+    remove_or_warn('docId')
+
+    # await collection.upsert("doc-a", {})
+    collection.upsert('doc-b', {})
+    collection.upsert('doc-c', {})
+    collection.upsert('doc-id', {})
+    collection.upsert('a-doc', {})
+
+    def txn_insert(ctx):
+        ctx.insert(collection, test_doc, 'hello')
+
+    try:
+        cluster.transactions.run(txn_insert)
+    except TransactionFailed as ex:
+        print(f'Transaction did not reach commit point.  Error: {ex}')
+    except TransactionCommitAmbiguous as ex:
+        print(f'Transaction possibly committed.  Error: {ex}')
+    
+    # tag::create[]
+    def txn_logic_ex(ctx  # type: AttemptContext
```

Comment:
this is what requires the TYPE_CHECKING etc. above?

Is this widely used in Python these days?

(The fact that you've not used it for the other examples, but only this one, make me wonder if it's really needed here? e.g. is there much benefit over just commenting `AttemptContext`?)

Reply:
The TYPE_CHECKING part is just a remnant of the coded example that Jared gave me, If you think it would be better to remove than I am happy to